drivers/pwm/pwm_led_esp32.c

```diff
@@ -0,0 +1,470 @@
+/*
+ * Copyright (c) 2017 Vitor Massaru Iha <vitor@massaru.org>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/* Include esp-idf headers first to avoid redefining BIT() macro */
+#include <esp_intr_alloc.h>
+#include <soc/dport_reg.h>
+#include <rom/gpio.h>
+#include <soc/gpio_sig_map.h>
+#include <soc/ledc_reg.h>
+
+#include <soc.h>
+#include <errno.h>
+#include <board.h>
+#include <pwm.h>
+#include <kernel.h>
+#include <gpio.h>
+#include <string.h>
+
+enum {
+	PWM_LED_ESP32_REF_TICK_FREQ,
+	PWM_LED_ESP32_APB_CLK_FREQ,
+};
+
+enum {
+	PWM_LED_ESP32_HIGH_SPEED,
+	PWM_LED_ESP32_LOW_SPEED
+};
+
+#define PWM_LED_ESP32_CHANNEL_NOT_USED -1
+
+/* XXX Move struct peripheral for other place ? */
+struct pwm_led_esp32_peripheral {
+	int clk;
+	int rst;
+};
+
+struct pwm_led_esp32_timer {
+	/* set timer counter bit number [10; 15] */
+	int bit_num;
+	int timer_sel;
+	int freq;
+};
+
+struct pwm_led_esp32_data {
+	/* pin-channel mapping */
+	int pin_ch[40];
+	/* channels used in high speed mode [1, 8] */
+	int hs_ch_used;
+	/* channels used in low speed mode [1, 8] */
+	int ls_ch_used;
+	/* GPIO pin on high speed channel */
+	int hs_ch_gpio[8];
+	/* GPIO pion on low speed channel */
+	int ls_ch_gpio[8];
+};
+
+struct pwm_led_esp32_gpio_speed_mode_range {
+	int lo;
+	int hi;
+};
+
+struct pwm_led_esp32_config {
+	/* We have 4 independent timers for high and low speed channels,
+	 * but for simplicity we will only use 1 per speed mode.
+	 */
+
+	/* Speed timer
+	 * 0: High speed timer
+	 * 1: Low speed timer
+	 */
+	struct pwm_led_esp32_timer timer_cfg[2];
+
+	struct pwm_led_esp32_gpio_speed_mode_range gpio_range_hs;
+	struct pwm_led_esp32_gpio_speed_mode_range gpio_range_ls;
+
+	const struct pwm_led_esp32_peripheral peripheral;
+};
+
+/* TODO: Remove these functions after this PR:
+ * https://github.com/zephyrproject-rtos/zephyr/pull/5113
+ */
+static inline void set_mask32(u32_t v, u32_t mem_addr)
+{
+	sys_write32(sys_read32(mem_addr) | v, mem_addr);
+}
+
+static inline void clear_mask32(u32_t v, u32_t mem_addr)
+{
+	sys_write32(sys_read32(mem_addr) & ~v, mem_addr);
+}
+
+static void esp32_enable_peripheral(
+	const struct pwm_led_esp32_peripheral *peripheral)
+{
+	set_mask32(peripheral->clk, DPORT_PERIP_CLK_EN_REG);
+	clear_mask32(peripheral->rst, DPORT_PERIP_RST_EN_REG);
+}
+
+static const char *esp32_get_gpio_for_pin(int pin)
+{
+	if (pin < 32) {
+#if defined(CONFIG_GPIO_ESP32_0)
+		return CONFIG_GPIO_ESP32_0_NAME;
+#else
+		return NULL;
+#endif /* CONFIG_GPIO_ESP32_0 */
+	}
+
+#if defined(CONFIG_GPIO_ESP32_1)
+	return CONFIG_GPIO_ESP32_1_NAME;
+#else
+	return NULL;
+#endif /* CONFIG_GPIO_ESP32_1 */
+}
+/* end Remove after PR 5113 */
+
+static void pwm_led_esp32_low_speed_update(int speed_mode, int channel)
+{
+	u32_t reg_addr;
+
+	if (speed_mode == PWM_LED_ESP32_LOW_SPEED) {
+		reg_addr = LEDC_LSCH0_CONF0_REG + 14*channel;
```

Comment:
Nitpick: `14*channel` should be written as `14 * channel`.

Also, some of the esp-idf macros for other peripherals will often have one that will calculate the register address given a parameter such as the channel. Do we have this for ledc?

Reply:
Unfortunately we don't have this for the ledc.

Reply:
I wrote these macros and fortunately I found many mistakes: I used the offset as decimal instead hexadecimal (ex.: 14 instead 0x14).